### PR TITLE
standalone Rrn ts encoder

### PR DESCRIPTION
**Pull request by torrmal**

### CODE REVIEW

File: lightwood/encoders/time_series/helpers/rnn_helpers.py [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+class DecoderRNNNumerical(nn.Module):
+    def __init__(self, hidden_size, output_size):
+        super(DecoderRNNNumerical, self).__init__()
+        self.hidden_size = hidden_size
+        self.gru = nn.GRU(output_size, hidden_size)
+        self.out = nn.Linear(hidden_size, output_size)
+
+
+    def forward(self, input, hidden):
```
Comment (George3d6) [inline lightwood/encoders/time_series/helpers/rnn_helpers.py:26]: Hmh, thinking about it, does it even make sense to have a DecoderRNN at all ? As in, wouldn't the "decoding" part be just the `Linear` layer of the EncoderRNN that we train.

File: lightwood/encoders/time_series/rnn.py [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+        :param as_list: if you want to return the information as lists
+        :param return_next_value:  if you want to return the next value in the time series too
+
+        :return:  either encoded_value or encoded_value, next_value
+        """
+        if not self._prepared:
+            raise Exception('You need to call "prepare_encoder" before calling "encode" or "decode".')
+
+        self._encoder.eval()
+        with torch.no_grad():
```
Comment (George3d6) [inline lightwood/encoders/time_series/rnn.py:83]: Ok, so... unrelated to this PR, but:

Initially I thought the `.eval` also stopped auto-grad from working. Reading a bit more on this it seems that the way you are doing it here is "correct" in that eval just puts the model in evaluation mode but doesn't disable autograd.

In this case, should be be using eval+no_grad for every instance when we are just using a model for predicting ? (It's currently not done, unrelated to this PR, but could yield a significant speed gain in many cases I'd wager)

File: lightwood/encoders/time_series/rnn.py [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+
+import torch
+import torch.nn as nn
+from torch import optim
+
+import numpy as np
+
+class RnnEncoder:
+
+    def __init__(self, encoded_vector_size=4, train_iters=75000, stop_on_error=0.8, learning_rate=0.01, is_target=False):
```
Comment (George3d6) [inline lightwood/encoders/time_series/rnn.py:12]: I'm not really sure it makes sense to have: `train_iters`, `stop_on_error` and `learning_rate` as part of the interface here, since we don't for the other encoders that require training.

It might make more sense to set them as public variables (without the `_`) and have them be overidden dynamically via the config as we do with the other encoder when we initialize the data source:

```
                encoder_attrs = config['encoder_attrs'] if 'encoder_attrs' in config else {}
                for attr in encoder_attrs:
                    if hasattr(encoder_instance, attr):
                        setattr(encoder_instance, attr, encoder_attrs[attr])
```

File: lightwood/encoders/time_series/helpers/rnn_helpers.py [hunk 4]
```diff
@@ -0,0 +1,54 @@
+
+
+
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+
+device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
```
Comment (George3d6) [inline lightwood/encoders/time_series/helpers/rnn_helpers.py:8]: This needs to be:

```
        device_str = "cuda" if CONFIG.USE_CUDA else "cpu"
        if CONFIG.USE_DEVICE is not None:
            device_str = CONFIG.USE_DEVICE
```

This is the default device setting behavior for everything else and it allows the user to specify cpu even if they have a gpu and to specify training on only a specific device (if they have multiple ones).

File: lightwood/encoders/time_series/rnn.py [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+                loss = 0
+
+                optimizer.zero_grad()
+                encoder_hidden = self._encoder.initHidden()
+                next_tensor = data_tensor[0]
+                for tensor_i in range(len(priming_data[j])-1):
+                    rand = np.random.randint(2)
+                    # teach from forward as well as from known tensor alteratively
+                    if rand == 1:
+                        next_tensor, encoder_hidden = self._encoder.forward(data_tensor[tensor_i] , encoder_hidden)
```
[the remark this answers is not in the document]
Comment (torrmal) [reply lightwood/encoders/time_series/rnn.py:51]: eyah if you can add that, please do

File: lightwood/encoders/time_series/helpers/rnn_helpers.py [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+class DecoderRNNNumerical(nn.Module):
+    def __init__(self, hidden_size, output_size):
+        super(DecoderRNNNumerical, self).__init__()
+        self.hidden_size = hidden_size
+        self.gru = nn.GRU(output_size, hidden_size)
+        self.out = nn.Linear(hidden_size, output_size)
+
+
+    def forward(self, input, hidden):
+        output = F.relu(input.float())
```
[the remark this answers is not in the document]
Comment (torrmal) [reply lightwood/encoders/time_series/helpers/rnn_helpers.py:27]: ? he is how you forward the input, not sure what you mean here

[the remark this answers is not in the document]
Comment (torrmal) [reply lightwood/encoders/time_series/helpers/rnn_helpers.py:27]: the graph you define in the forward on the init you define the elements of the graph

File: lightwood/encoders/time_series/helpers/rnn_helpers.py [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+
+
+
+import numpy as np
+
+def tensor_from_series(series):
+    return torch.tensor(series, dtype=torch.float, device=device).view(-1, 1, 1, 1).float()
+
+
+class DecoderRNNNumerical(nn.Module):
```
[the remark this answers is not in the document]
Comment (torrmal) [reply lightwood/encoders/time_series/helpers/rnn_helpers.py:18]: this is so you can take time series as input, which is the use cases that we have, we need to evaluate ts as output before we implement the decode

File: lightwood/encoders/time_series/helpers/rnn_helpers.py (hunk 1 above)
Comment (torrmal) [reply lightwood/encoders/time_series/helpers/rnn_helpers.py:26]: it depends on what we define for ts as target

Comment (torrmal) [reply lightwood/encoders/time_series/helpers/rnn_helpers.py:26]: separate PR, see previous questions

File: lightwood/encoders/time_series/rnn.py (hunk 2 above)
Comment (torrmal) [reply lightwood/encoders/time_series/rnn.py:83]: sure, do a separate issue, and add that to the other parts needed

File: lightwood/encoders/time_series/rnn.py (hunk 3 above)
Comment (torrmal) [reply lightwood/encoders/time_series/rnn.py:12]: they are public variables, and you can set them via params on config, as such its ok to have these params here so the encoder can be used standalone, to facilitate moving to a factory pattern, separate PR for that

File: lightwood/encoders/time_series/helpers/rnn_helpers.py (hunk 4 above)
Comment (torrmal) [reply lightwood/encoders/time_series/helpers/rnn_helpers.py:8]: good point

File: lightwood/encoders/time_series/rnn.py [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+                    raise Exception('get_next_count must be greater than 0')
+
+                hidden = None
+                vector = val
+
+                next_i = []
+
+                for j in range(get_next_count):
+
+                    hidden, next_reading = self.encode_one(vector, initial_hidden=hidden, as_list=True, return_next_value=True)
```
[the remark this answers is not in the document]
Comment (torrmal) [reply lightwood/encoders/time_series/rnn.py:137]: I like the list, it lets you test, and its optional, it will come handy to test when we build the decoder

File: lightwood/encoders/time_series/rnn.py [hunk 9]
```diff
@@ -0,0 +1,178 @@
...
+
+                loss = loss
+                average_loss += int(loss)
+                loss.backward()
+                optimizer.step()
+            average_loss = average_loss/total
+            if average_loss < self._stop_on_error:
+                break
+            if feedback_hoop_function is not None:
+                feedback_hoop_function("epoch [{epoch_n}/{total}] average_loss = {average_loss}".format(average_loss=average_loss, epoch_n=i+1, total=self._train_iters))
```
[the remark this answers is not in the document]
Comment (torrmal) [reply lightwood/encoders/time_series/rnn.py:65]: if you want you can do the gym on a separate PR, after we merge this one.

File: lightwood/encoders/time_series/helpers/rnn_helpers.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (George3d6) [reply lightwood/encoders/time_series/helpers/rnn_helpers.py:27]: I mean to say we can add: https://pytorch.org/docs/stable/nn.html#torch.nn.ReLU as part of the graph, that's how we do it in the default net and in the selfaware (and similar thing for the losses, we initialize them as objects even though we could just as well use them as functions)

[the remark this answers is not in the document]
Comment (George3d6) [reply lightwood/encoders/time_series/helpers/rnn_helpers.py:27]: It would be the same functionality, it's just that it would be more consistent with the rest of the code.

File: lightwood/encoders/time_series/helpers/rnn_helpers.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (George3d6) [reply lightwood/encoders/time_series/helpers/rnn_helpers.py:18]: Alright, well, we can discuss ts outputs separately from this PR (if it's even something we'd want to implement ?)

File: lightwood/encoders/time_series/rnn.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (George3d6) [reply lightwood/encoders/time_series/rnn.py:137]: Fair enough, let's leave the argument but not use the list inside the encode function itself in that case (to speed things up).

File: lightwood/encoders/time_series/rnn.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (George3d6) [reply lightwood/encoders/time_series/rnn.py:65]: Hmh, it's fairly easy to add, I can add it together with the other change (the fixed for training on the GPU and transfering from gpu to cpu and vice versa).

Unless you want me not to I'd rather add it on (Monday ?) when I make those other changes.

File: lightwood/encoders/time_series/rnn.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (George3d6) [reply lightwood/encoders/time_series/rnn.py:51]: Will do and will create an issue for adding the .to function to the other encoders.

File: lightwood/encoders/time_series/rnn.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (torrmal) [reply lightwood/encoders/time_series/rnn.py:65]: yes, lets move it to a separate PR

File: lightwood/encoders/time_series/rnn.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (torrmal) [reply lightwood/encoders/time_series/rnn.py:137]: the thing is that RNNs are interesting in the sense that  since the usually run on a row by row basis unless we knew that all the dequencences were the same length we cannot do batches, still you are going to have to do append calls, not sure if tolist() is any more expensive than the append to tensor operations, but feel free to give it a crack if you rather do those operations on the tensor space.

File: lightwood/encoders/time_series/helpers/rnn_helpers.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (torrmal) [reply lightwood/encoders/time_series/helpers/rnn_helpers.py:27]: still not sure, what you mean, feel free to modify and and I will review

**Comment by George3d6** [comment]
* Added support for training both on CPU and GPU
* Added support for the string timeseries format (1 2 3 4 5...etc), our other TS encoeers support both space-separated strings with numbers and arrays, so I believe we should support both.

Other than that just some minor cleaning up of the tests and the code. Will make another PR for the bigger changes (e.g. switching to training with the Gym)
